### DISCUSSION TITLE
Reading pulse status

### DISCUSSION THREAD

**Discussion by VagnHovman** [arendst/Tasmota#14897] [Q&A] [answered]
I have made a HW interface between my garage gate controller and an ESP D1 mini. I can now operate UP, STOP, DOWN and TOGGLE. It works just fine. Now I would like to be able to read the status of the gate. There is only one status LED which I have connected to an input pin on the ESP D1. 
LED status:
OFF = gate closed,
ON = gate stopped but not closed,
Flashing 1 per sec. =  gate moving up
Flashing 0.5 per sec. = gate moving down.
I would like to receive a mqtt message every time the status changes e.g. when I send a DOWN command I would like to receive "Moving down" and there after "Closed". I can read all ON and OFFs but what about all these flashes. I have tried with rules and timers, but it seems like the rules run in parallel and suddenly everything is out of sync. Does anyone have a suggestion for a solution?

**Comment by barbudor** [comment, marked as answer]
You could test using the GPIO as "Counter" and configuring it as a timer with `CounterType 1`
If not, maybe using the counter mode `CounterType 0` and using 2 rules
- one to pickup the cunter value in a variable `on counter#c1 do var1 %value% endon`
- another one on a rule timer, like every 5 or 10 seconds to check the value and reset the counter

**Reply by stefanbode** [reply]
Have you tried to use the shutter driver of tasmota. This will send you all the required information on mqtt. Check the documentation. There are many modes how to operate

**Reply by VagnHovman** [reply]
Hi Stefanbode, thanks for your suggestion. However it seems that the shutter driver is for controlling a shutter. What I need is to read a status based of a pulse rate, this I can not find among the many shutter features.

**Reply by VagnHovman** [reply]
Hi barbudor, thanks for your suggestion. In theory, your proposal is really good, but in practice it does not work so well. It seems that the rules are settled too slowly, so the solution does not become stable. As I wrote in my question it seems like these rules and events are running in parallel. Some commands are executed quickly others very slowly.

**Reply by sfromis** [reply]
While rules are event-driven, and not some sort of sequence, it will indeed act as if running in parallel, but that "only" means that you should consider that when creating rules. One potential way of detecting "pulsing" would be to config the input as a button/switch, select appropriate modes/options to let rules handle this. If you, say, config as a button, and create a rule to act on "button pushed", that rule could start a ruletimer for a period longer than the blink time. Each blink would thus restart that timer. If the timer expires, that means "no longer blinking".

**Reply by VagnHovman** [reply]
Hi Sfromis, that's exactly what I started doing and it also works in isolation. The rule sends "Moving" when pulses arrive and when the pulses stop, "Closed" or "No closed" is sent. It seems stable. The problem arises when I try to count the number of pulses in a certain period to determine if it is 1 or 2 pulses per second. Counting the pulses goes fine, ruletimer period is fine, but the rules/events that control these rules start very randomly. Therefore, I asked if there was another solution.

**Reply by sfromis** [reply]
You could have two detection rules+timers running. As long as the faster one counts pulses, disregard the slower one also counting.
